[opt](partial update) use a separate config to control the behavior of newly inserted rows in partial update

bobhan1 opened the pull request:
Currently, Doris use strict mode to decide if newly inserted rows should be appended or report an error in partial update, which is hard to use. This PR add a new session variable and load property `partial_update_new_key_behavior` to control the behavior of newly inserted rows in partial update.
`partial_update_new_key_behavior` has ~three~ two options:
- `APPEND`: append the newly inserted rows
- ~`IGNORE`: delete the newly inserted rows silently(will not be taken into filtered rows)~
- `ERROR`: report error if meet newly inserted rows, and the error msg will contains one row's keys which is not in table.

---
The reason for not supporting `IGNORE` mode: To support `IGNORE` mode, we need to add delete sign for newly inserted rows in partial update to delete them rather than use delete bitmap mark to delete them because compaction will not use delete bitmap when reading data. Also, we need to record the rows whose delete sign is added by us in this situation for resolving conflicts in publish phase to avoid wrongly delete the rows if there are another concurrent load insert some of these rows successfully. This increases code complexity and is error-prone.

Doc: https://github.com/apache/doris-website/pull/2472

bobhan1 commented:
run buildall

bobhan1 commented:
run buildall

zhannngchen commented on be/src/olap/rowset/segment_v2/segment_writer.cpp:
```diff
@@ -475,16 +475,15 @@ Status SegmentWriter::probe_key_for_mow(
     if (st.is<KEY_NOT_FOUND>()) {
-        if (_opts.rowset_ctx->partial_update_info->is_strict_mode) {
-            ++stats.num_rows_filtered;
+        auto ignore_cb = [&]() {
```
The `ignore` operation will not work after compaction, compaction will ignore delete bitmap, and keep the latest key, unless you add a delete sign to such row.
But adding delete sign might have some issue if user specified sequence column.

zhannngchen replied:
Maybe we need to consider to remove IGNORE option
The IGNORE option works exactly like the reverted PR for `INSERT IGNORE`

bobhan1 commented:
run buildall

zhannngchen reviewed (dismissed):
LGTM

bobhan1 committed:
fix

fix case

fix some case under cluster key dir

fix case

tmp

update

remove strict mode partial update doc case

print duplidate key in status error msg in ERROR mode

fix nereids stream load.

fix show var

fix and suport broker load

fix

bobhan1 commented:
run buildall

bobhan1 commented:
run buildall

bobhan1 commented:
run buildall

bobhan1 commented:
run buildall

bobhan1 commented:
run buildall

bobhan1 commented:
run cloudut

bobhan1 commented:
run buildall

bobhan1 commented:
run buildall

bobhan1 commented:
run buildall

bobhan1 commented:
run buildall

bobhan1 commented:
run buildall

hello-stephen commented:
# BE UT Coverage Report
Increment line coverage `6.56% (8/122)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/0770029d0625e138ecb5e966aeef48e13bd0b22a_0770029d0625e138ecb5e966aeef48e13bd0b22a/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/0770029d0625e138ecb5e966aeef48e13bd0b22a_0770029d0625e138ecb5e966aeef48e13bd0b22a/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 54.67% (14739/26962) |
| Line Coverage     | 43.73% (129068/295129) |
| Region Coverage   | 42.46% (65889/155167) |
| Branch Coverage   | 37.03% (33200/89668) |

bobhan1 commented:
run feut

bobhan1 commented:
run buildall

hello-stephen commented:
# BE UT Coverage Report
Increment line coverage `6.56% (8/122)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/0770029d0625e138ecb5e966aeef48e13bd0b22a_0770029d0625e138ecb5e966aeef48e13bd0b22a/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/0770029d0625e138ecb5e966aeef48e13bd0b22a_0770029d0625e138ecb5e966aeef48e13bd0b22a/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 54.67% (14740/26964) |
| Line Coverage     | 43.73% (129090/295170) |
| Region Coverage   | 42.47% (65903/155182) |
| Branch Coverage   | 37.03% (33209/89678) |

bobhan1 commented:
run buildall

bobhan1 commented:
run buildall

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `77.24% (95/123)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/41232_4265c237e47b71c684a4b6afad53332a03d6761b_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/41232_4265c237e47b71c684a4b6afad53332a03d6761b_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 80.16% (21204/26451) |
| Line Coverage     | 74.25% (218801/294700) |
| Region Coverage   | 72.54% (131746/181615) |
| Branch Coverage   | 65.74% (67234/102266) |

bobhan1 commented:
run buildall

bobhan1 commented:
run buildall

starocean999 commented on fe/fe-core/src/main/java/org/apache/doris/planner/StreamLoadPlanner.java:
```diff
@@ -328,6 +328,9 @@ public TPipelineFragmentParams plan(TUniqueId loadId, int fragmentInstanceIdInde
         olapTableSink.init(loadId, taskInfo.getTxnId(), db.getId(), timeout, taskInfo.getSendBatchParallelism(),
                 taskInfo.isLoadToSingleTablet(), taskInfo.isStrictMode(), txnTimeout);
         olapTableSink.setPartialUpdateInfo(uniquekeyUpdateMode, partialUpdateInputColumns);
+        if (uniquekeyUpdateMode != TUniqueKeyUpdateMode.UPSERT) {
```
the old StreamLoadPlanner is replaced by NereidsStreamLoadPlanner and will be deleted soon. Is there any unexpected dependence, so that have to do these changes in old code?

hello-stephen commented:
# Cloud UT Coverage Report
Increment line coverage ` ` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/0e759f84d2a7baf1b8ed63c7b1a44bc941fd2159_0e759f84d2a7baf1b8ed63c7b1a44bc941fd2159_cloud/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/0e759f84d2a7baf1b8ed63c7b1a44bc941fd2159_0e759f84d2a7baf1b8ed63c7b1a44bc941fd2159_cloud/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 83.15% (1105/1329) |
| Line Coverage     | 66.05% (18610/28175) |
| Region Coverage   | 65.64% (9267/14117) |
| Branch Coverage   | 55.33% (4981/9002) |

bobhan1 commented:
run buildall

hello-stephen commented:
# BE UT Coverage Report
Increment line coverage `7.14% (8/112)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/b55b6bbd3246648db15b7e2bc9bab298aa1e6ad7_b55b6bbd3246648db15b7e2bc9bab298aa1e6ad7/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/b55b6bbd3246648db15b7e2bc9bab298aa1e6ad7_b55b6bbd3246648db15b7e2bc9bab298aa1e6ad7/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 55.80% (14900/26701) |
| Line Coverage     | 44.59% (131877/295760) |
| Region Coverage   | 43.64% (66289/151895) |
| Branch Coverage   | 38.25% (33968/88798) |

bobhan1 commented:
run feut

bobhan1 commented:
run buildall

hello-stephen commented:
# BE UT Coverage Report
Increment line coverage `7.14% (8/112)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/7f3198a312bfd9c6d86759b9839428ab2d434054_7f3198a312bfd9c6d86759b9839428ab2d434054/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/7f3198a312bfd9c6d86759b9839428ab2d434054_7f3198a312bfd9c6d86759b9839428ab2d434054/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 55.80% (14900/26701) |
| Line Coverage     | 44.59% (131883/295760) |
| Region Coverage   | 43.64% (66294/151895) |
| Branch Coverage   | 38.25% (33969/88798) |

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `74.34% (84/113)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/41232_7f3198a312bfd9c6d86759b9839428ab2d434054_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/41232_7f3198a312bfd9c6d86759b9839428ab2d434054_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 79.40% (20856/26267) |
| Line Coverage     | 72.65% (214820/295696) |
| Region Coverage   | 70.88% (126448/178391) |
| Branch Coverage   | 64.60% (65503/101404) |

bobhan1 commented:
run feut

bobhan1 commented:
run buildall

hello-stephen commented:
# FE UT Coverage Report
Increment line coverage `  54.22% (45/83)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/41232_4c1752e9653e5fd0cb9a42612ebc56c048abe9fd/fe_increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/41232_4c1752e9653e5fd0cb9a42612ebc56c048abe9fd/fe_report/index.html)

zhannngchen reviewed (dismissed):
LGTM

bobhan1 committed:
change name to partial_update_new_key_behavior

bobhan1 commented:
run buildall

hello-stephen commented:
# Cloud UT Coverage Report
Increment line coverage ` ` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/e60c665c3727ecedb3f6c5d1421e96c1da5bd56a_e60c665c3727ecedb3f6c5d1421e96c1da5bd56a_cloud/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/e60c665c3727ecedb3f6c5d1421e96c1da5bd56a_e60c665c3727ecedb3f6c5d1421e96c1da5bd56a_cloud/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 83.33% (1120/1344) |
| Line Coverage     | 66.80% (19310/28909) |
| Region Coverage   | 66.57% (9579/14390) |
| Branch Coverage   | 56.49% (5206/9216) |

zhannngchen reviewed (dismissed):
LGTM

hello-stephen commented:
# FE UT Coverage Report
Increment line coverage `  56.96% (45/79)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/41232_e60c665c3727ecedb3f6c5d1421e96c1da5bd56a/fe_increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/41232_e60c665c3727ecedb3f6c5d1421e96c1da5bd56a/fe_report/index.html)

hello-stephen commented:
# BE UT Coverage Report
Increment line coverage `7.08% (8/113)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/48684603b1e390261b8031d255492f9923b6191a_48684603b1e390261b8031d255492f9923b6191a/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/48684603b1e390261b8031d255492f9923b6191a_48684603b1e390261b8031d255492f9923b6191a/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 56.40% (15075/26727) |
| Line Coverage     | 45.16% (134783/298438) |
| Region Coverage   | 44.28% (67830/153181) |
| Branch Coverage   | 38.85% (34816/89614) |

bobhan1 commented:
run external

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `73.68% (84/114)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/41232_48684603b1e390261b8031d255492f9923b6191a_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/41232_48684603b1e390261b8031d255492f9923b6191a_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 79.69% (20961/26304) |
| Line Coverage     | 72.63% (216644/298280) |
| Region Coverage   | 70.91% (127735/180145) |
| Branch Coverage   | 64.63% (66188/102416) |

bobhan1 commented:
run buildall

hello-stephen commented:
# Cloud UT Coverage Report
Increment line coverage ` ` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/719746f67c9b70df82d495ff3cf0f02cb459d208_719746f67c9b70df82d495ff3cf0f02cb459d208_cloud/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/719746f67c9b70df82d495ff3cf0f02cb459d208_719746f67c9b70df82d495ff3cf0f02cb459d208_cloud/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 83.33% (1120/1344) |
| Line Coverage     | 66.85% (19325/28909) |
| Region Coverage   | 66.59% (9583/14390) |
| Branch Coverage   | 56.54% (5211/9216) |

bobhan1 commented:
run p0

bobhan1 commented:
run performance

bobhan1 commented:
run external

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `8.77% (10/114)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/41232_719746f67c9b70df82d495ff3cf0f02cb459d208_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/41232_719746f67c9b70df82d495ff3cf0f02cb459d208_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 60.98% (16034/26293) |
| Line Coverage     | 50.48% (150582/298286) |
| Region Coverage   | 47.82% (86063/179968) |
| Branch Coverage   | 41.33% (42261/102264) |

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `8.77% (10/114)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/41232_719746f67c9b70df82d495ff3cf0f02cb459d208_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/41232_719746f67c9b70df82d495ff3cf0f02cb459d208_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 61.83% (16258/26293) |
| Line Coverage     | 51.28% (152965/298286) |
| Region Coverage   | 48.79% (87811/179968) |
| Branch Coverage   | 42.26% (43214/102264) |

dataroaring reviewed (approved):
LGTM
